### DISCUSSION TITLE
How to use aggregate api with nexus/schema

### DISCUSSION THREAD

**Discussion by 45592858** [prisma/prisma#4353] [Q&A] [answered]
I tried with example [graphql-express](https://github.com/prisma/prisma-examples/tree/latest/typescript/graphql-express), add "money" field in User model, and try aggregate api, but show below error.

Error:
- Missing type AggregateUser, did you forget to import a type to the root query?
    at Object.assertNoMissingTypes (D:\Temp\temp\try\graphql-express\node_modules\@nexus\schema\src\utils.ts:323:11)
    at Object.makeSchema (D:\Temp\temp\try\graphql-express\node_modules\@nexus\schema\src\builder.ts:1616:3)
    at Object.<anonymous> (D:\Temp\temp\try\graphql-express\src\schema.ts:116:23)
    at Module._compile (internal/modules/cjs/loader.js:955:30)
    at Module._compile (D:\Temp\temp\try\graphql-express\node_modules\source-map-support\source-map-support.js:547:25)
    at Module.m._compile (C:\Users\GZD53E~1.YAN\AppData\Local\Temp\ts-node-dev-hook-2821629441210556.js:60:25)
    at Module._extensions..js (internal/modules/cjs/loader.js:991:10)
    at require.extensions.<computed> (C:\Users\GZD53E~1.YAN\AppData\Local\Temp\ts-node-dev-hook-2821629441210556.js:62:14)
    at Object.nodeDevHook [as .ts] (D:\Temp\temp\try\graphql-express\node_modules\ts-node-dev\lib\hook.js:61:7)
    at Module.load (internal/modules/cjs/loader.js:811:32)
[ERROR] 10:48:52 Error:
- Missing type AggregateUser, did you forget to import a type to the root query?

------------------------------------------
Source Code:
------------------------------------------

```ts
import { intArg, makeSchema, objectType, stringArg } from '@nexus/schema'
import { nexusPrisma } from 'nexus-plugin-prisma'

const User = objectType({
  name: 'User',
  definition(t) {
    t.model.id()
    t.model.name()
    t.model.money()
    t.model.email()
    t.model.posts({
      pagination: false,
    })
  },
})

const Post = objectType({
  name: 'Post',
  definition(t) {
    t.model.id()
    t.model.title()
    t.model.content()
    t.model.published()
    t.model.author()
  },
})

const Query = objectType({
  name: 'Query',
  definition(t) {
    t.crud.post()

    t.list.field('feed', {
      type: 'Post',
      resolve: (_parent, _args, ctx) => {
        return ctx.prisma.post.findMany({
          where: { published: true },
        })
      },
    })

    t.list.field('filterPosts', {
      type: 'Post',
      args: {
        searchString: stringArg({ nullable: true }),
      },
      resolve: (_, { searchString }, ctx) => {
        return ctx.prisma.post.findMany({
          where: {
            OR: [
              { title: { contains: searchString } },
              { content: { contains: searchString } },
            ],
          },
        })
      },
    })

    t.field('userMoneySum', {
      type: 'AggregateUser',
      resolve: (_parent, _args, ctx) => {
        return ctx.prisma.user.aggregate({
          sum: {
            money: true,
          }
        })
      },
    })

  },
})

const Mutation = objectType({
  name: 'Mutation',
  definition(t) {
    t.crud.createOneUser({ alias: 'signupUser' })
    t.crud.deleteOnePost()

    t.field('createDraft', {
      type: 'Post',
      args: {
        title: stringArg({ nullable: false }),
        content: stringArg(),
        authorEmail: stringArg({ nullable: false }),
      },
      resolve: (_, { title, content, authorEmail }, ctx) => {
        return ctx.prisma.post.create({
          data: {
            title,
            content,
            published: false,
            author: {
              connect: { email: authorEmail },
            },
          },
        })
      },
    })

    t.field('publish', {
      type: 'Post',
      nullable: true,
      args: {
        id: intArg(),
      },
      resolve: (_, { id }, ctx) => {
        return ctx.prisma.post.update({
          where: { id: Number(id) },
          data: { published: true },
        })
      },
    })
  },
})

export const schema = makeSchema({
  types: [Query, Mutation, Post, User],
  plugins: [nexusPrisma({ experimentalCRUD: true })],
  outputs: {
    schema: __dirname + '/../schema.graphql',
    typegen: __dirname + '/generated/nexus.ts',
  },
  typegenAutoConfig: {
    contextType: 'Context.Context',
    sources: [
      {
        source: '@prisma/client',
        alias: 'prisma',
      },
      {
        source: require.resolve('./context'),
        alias: 'Context',
      },
    ],
  },
})
```

**Comment by ryands17** [comment, marked as answer]
@45592858 
You would need to create a new type named `AggregateUser` and update the resolver:

```ts
t.field('userMoneySum', {
      type: 'AggregateUser',
      resolve: async (_parent, _args, ctx) => {
        return (await prisma.user.aggregate({ sum: { money: true } })).sum
      },
    })
}

const AggregateUser = objectType({
  name: 'AggregateUser,
  definition(t) {
    t.float('money')
  },
})
```

Finally add the `AggregateUser` to the types:

```ts
types: [Query, Mutation, Post, User, AggregateUser],
```

**Comment by 45592858** [comment]
but the AggregateUser seems already auto generated in .prisma/client/index.js.
and it also mentions on [prisma aggregations](https://www.prisma.io/docs/concepts/components/prisma-client/aggregations)

**Reply by ryands17** [reply]
That is the type generated by Prisma which is different than the GraphQL `type`. This type needs to be created that will be available in your `schema.graphql` and that will be used as a type.

Prisma types are just TypeScript types, while Nexus generates GraphQL types for you. [This](https://graphql.org/learn/schema/) is an article that explains on GraphQL types.

**Reply by 45592858** [reply]
Great, thanks you @ryands17 very much, it is work now.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
